### PR TITLE
Add snow particles that fall from the sky

### LINKED ISSUE #1: Snowing effect when player is standing on snow
Here's a possibly lightweight implementation:

- Check if a player is standing on snowy nodes (and maybe also if player is standing directly under the sky).
- If yes, attach a particle spawner to the player that creates the effect of snow.

### PR DESCRIPTION
**Pull request by LoneWolfHT**
Closes #1

### CODE REVIEW

File: snow/init.lua
Comment (ClobberXD) [inline snow/init.lua]: `ctf_match.register_on_new_match` would be better here. Why snow only after buildtime?

Comment (LoneWolfHT) [reply snow/init.lua]: It was due to my original attempts to find the map bounds, I was using the flag positions. Can change that now though...

Comment (ClobberXD) [reply snow/init.lua]: There's the issue of the map using the `on_new_match` too, so there's the possibility of a race condition. But as this mod depends on `ctf_map`, the latter will be executed first, and would subsequently register its `on_new_match` first.

File: snow/init.lua
Comment (ClobberXD) [inline snow/init.lua]: Why are the x and z coords decremented too?

Comment (LoneWolfHT) [reply snow/init.lua]: I'm assuming without that it would set them to spawn inside the barriers at the edges of the map?
I've removed this in my upcoming update anyway since from what I've seen the map coordinates can vary (both positive and negative) and it probably wouldn't be worthwhile to add the calculations to make it independent of that

File: snow/init.lua
Comment (LoneWolfHT) [reply snow/init.lua]: Changed

File: snow/init.lua
Comment (LoneWolfHT) [reply snow/init.lua]: Removed

File: snow/init.lua
Comment (ClobberXD) [inline snow/init.lua]: How does this part remove the particlespawner when the match is over?

Comment (LoneWolfHT) [reply snow/init.lua]: See line `8`

Comment (ClobberXD) [reply snow/init.lua]: Oh I misread, I thought it was `particlespawner == nil`, sorry.

File: snow/init.lua
Comment (ClobberXD) [inline snow/init.lua]: Spacing required around the `/` symbols.

File: snow/init.lua
Comment (ClobberXD) [inline snow/init.lua]: Spacing required around `=` symbols

**Comment by ClobberXD** [comment]
- Snow underground and indoors looks really weird
- Particlespawners still do get unloaded in maps where the bases are pretty low, like Caverns.

**Comment by ClobberXD** [comment]
> AFAIK particles are clientside. So it's the client's unloaded mapblocks that are causing the problem. The particles fall fine for me with 300 viewing range, but when I set my viewing range to 20 they don't fall since the particles collide with the unloaded mapblock's 'ignore' due to it being walkable.

I agree that particles are client-side, but reducing viewing range doesn't automatically spawn ignore nodes outside the visible range. The reason the particles don't fall down is that they aren't even spawned in the first place, as the mapblocks that "host" the particlespawner are unloaded. Mapblocks being loaded is controlled clientside via the `number_of_mapblocks_to_store_in_memory` or a similarly named setting, and also server-side, via the `block_send_radius` or a similarly-named setting. Some clients have the former set to a high value (the default value is pretty decent by itself, IIRC), and can keep the block loaded even if they're pretty far away, having not reached the set limit for blocks in memory. Ergo, `view_range` does not matter as much as these other settings.

**Comment by LoneWolfHT** [comment]
Any suggestions on how to get it working? If I read your message right you just said it's not possible without changing a client-side option?

**Comment by ClobberXD** [comment]
> If I read your message right you just said it's not possible without changing a client-side option?

From what I know, that's not possible, yes.

> Any suggestions on how to get it working?

- Attach particle-spawners to each player, and spawn snow around them. The number of particle-spawners won't be an issue because particle-spawners are completely client-side, just like individual particles.

- [Optional] Using a globalstep or recursive `minetest.after`, check for the presence of non-air nodes up to around 50-100 nodes (or the top of the schem, whichever is lower) above each player, preferably using LVMs for the sake of performance. If there are no obstructions, continue spawning snow, else stop spawning snow. Maybe repeat this check every 5 seconds?

**Comment by LoneWolfHT** [comment]
Yeah, none of that sounds very good. Maybe once #3 is merged I could have snow spawn 15 nodes above `default:snow` and `winterize:ice` with collision detection on

**Comment by ClobberXD** [comment]
True, there's no best way here, I'm afraid...

> I could have snow spawn 15 nodes above `default:snow` and `winterize:ice` with collision detection on

One particle spawner for each node on the surface? That'd be bad for performance. Also, 15 nodes is too less. What if a player is higher than 15 nodes from the surface? Snow would spawn below them, and would end up breaking the immersion.

**Comment by LoneWolfHT** [comment]
Hmmm, not sure why I did that. I hadn't tested. Snow still doesn't fall 
image
(This is exactly what I got from the last method btw. The snow is spawning it just isn't falling or colliding with anything it seems)

**Comment by LoneWolfHT** [comment]
30 nodes above a player is the best I can do to get the snow to fall with 20 viewing range. I'm going to try lowering where the particlespawner snow spawns from. Since most maps have quite a bit of space at the top.
In the end I think we'll just need to ask players to turn their viewing range up and turning it back down to normal once the snow starts spawning

**Comment by LoneWolfHT** [comment]
I've also created an issue https://github.com/minetest/minetest/issues/9127

**Comment by ClobberXD** [comment]
@LoneWolfHT Give the idea in https://github.com/MT-CTF/seasonal_xmas/pull/2#issuecomment-554718744 a shot, if you haven't already.

**Comment by LoneWolfHT** [comment]
@ClobberXD It doesn't work once you start a new match though. No idea why, you can take a look if you want

**Comment by ClobberXD** [comment]
Strange; the code seems fine at first glance...

On a side-note, you might want to define [`playername`](https://github.com/minetest/minetest/blob/4b6bff46e14c6215828da5ca9ad2cb79aa517a6e/doc/lua_api.txt#L7424) for each particle spawner, otherwise when two players are PvPing, there'd be two times as much snowfall around them. Oh, and there'd be a blizzard in the team-base as long as there are players in it. ;)

**Comment by LoneWolfHT** [comment]
> On a side-note, you might want to define [`playername`](https://github.com/minetest/minetest/blob/4b6bff46e14c6215828da5ca9ad2cb79aa517a6e/doc/lua_api.txt#L7424) for each particle spawner, otherwise when two players are PvPing, there'd be two times as much snowfall around them. Oh, and there'd be a blizzard in the team-base as long as there are players in it. ;)

Yep, it's in there. Just hiding lol

**Comment by ClobberXD** [comment]
Oh nice, I didn't notice that line before.

**Comment by ClobberXD** [comment]
Tested, and I can confirm that this doesn't work. I suspect this might be due to the fact we the player objects are invisible to ourselves, and attaching anything to our own player objects would make them invisible too.

However, I'll mess around with the code to see if I can make something happen.

**Comment by LoneWolfHT** [comment]
@ClobberXD Turns out it's an engine bug https://github.com/minetest/minetest/issues/9152

**Comment by LoneWolfHT** [comment]
Shall I look into making an ABM that will spawn a snow particles above the snow 'slab's every once and a while?

**Comment by ClobberXD** [comment]
While ABMs would work, I'm afraid they'll cause a good bit of performance hit, especially when running on such massive regions. Feel free to give it a shot though, and do share your results here. I'd also recommend you take a look at how weather mods implement rain and snow.

> Turns out it's an engine bug minetest/minetest#9152

Yes, that's exactly what I said too :)

> > I suspect this might be due to the fact we the player objects are invisible to ourselves, and attaching anything to our own player objects would make them invisible too.

**Comment by LoneWolfHT** [comment]
@ClobberXD and @Thomas--S: Ready for review

**Comment by ClobberXD** [comment]
Spawning individual particles would work, but they have a much greater network performance penalty, as the server needs to send `TOCLIENT_ADD_PARTICLE` for each and every single particle. Assuming there are 30 players online, the server has to send 210 (30 * 7) packets every 0.7 seconds. i.e. **300 packets/s (210 / 0.7)**. While the clients don't have any extra load (except for handling a huge number of packets, one for each particle), and render a similar amount of particles as before, but the server takes a huge performance hit, and for this reason, I'm not happy with the current implementation. \*sigh\* I really wish attached particlespawners worked... :/

**Comment by LoneWolfHT** [comment]
I guess we just put this on hold until it's fixed then? Because if this is laggy then ABMs probably will be too

**Comment by ClobberXD** [comment]
There could be other ways to do this that we can't think of, at the moment. Have you taken a look at how weather mods implement precipitation? I'm starting to think that those mods too might use a technique that puts a lot of load on the server, and would hence only be used in singleplayer. But we could give them a shot, nonetheless. If that fails, then I'm afraid we'd have to put this beautiful feature on hold.

**Comment by LoneWolfHT** [comment]
checked paramat's mod. He uses something similar to my method

**Comment by ClobberXD** [comment]
Ahh, I see. We may be out of options then. Snowfall would've been the highlight of this modpack... :(

**Comment by Thomas--S** [comment]
I found a solution based on LoneWolf's approach but with far less network packets (as far as I can see).
Se LoneWolfHT/seasonal_xmas#1

**Comment by ClobberXD** [comment]
Nice; using a particlespawner is the logical thing to do in these scenarios. Will test within a couple of days.

**Review by Thomas--S — APPROVED** [review]

**Comment by ClobberXD** [comment]
Indentation needs to be fixed. PR can be merged without my approval.

**Comment by rubenwardy** [comment]
Will it snow in maps like caverns? :D

**Review by rubenwardy — APPROVED** [review]

**Comment by LoneWolfHT** [comment]
@rubenwardy I noticed a little snow in the `Kingdom in The Nether` map. If you want I could use the `is_snow_obstructed()` check from my winterize PR to prevent that.
Snow will probably appear in caverns too but only when you have a lot of air above you

**Pull request merged**
